pkg/grpc/server.go

```diff
@@ -124,7 +124,9 @@ func getDefaultMaxAgeDuration() *time.Duration {
 // StartNonBlocking starts a new server in a goroutine.
 func (s *server) StartNonBlocking() error {
 	var listeners []net.Listener
-	if s.config.UnixDomainSocket != "" && s.kind == apiServer {
+	if s.config.listener != nil {
+		listeners = append(listeners, s.config.listener)
+	} else if s.config.UnixDomainSocket != "" && s.kind == apiServer {
```

Comment:
is there a way to implement this without introducing the if else clause here?

Reply:
you can create a helper function `getListeners` and just return `listeners` slice at each if step

Reply:
That would move the if-else clause from here into the helper function.

Reply:
in some way but without the deeper nesting
```
func getListeners() ([]net.Listener, error) {
	if s.config.listener != nil {
		return []net.Listener{s.config.listener}, nil
	if s.config.UnixDomainSocket != "" && s.kind == apiServer {
		socket := fmt.Sprintf("%s/dapr-%s-grpc.socket", s.config.UnixDomainSocket, s.config.AppID)
		l, err := net.Listen("unix", socket)
		if err != nil {
			return nil, err
		}
		s.logger.Infof("gRPC server listening on UNIX socket: %s", socket)
		return []net.Listener{l}, nil
	listeners := make([]net.Listener,0,len(s.config.APIListenAddresses))
	for _, apiListenAddress := range s.config.APIListenAddresses {
		addr := apiListenAddress + ":" + strconv.Itoa(s.config.Port)
		l, err := net.Listen("tcp", addr)
		if err != nil {
			s.logger.Debugf("Failed to listen for gRPC server on TCP address %s with error: %v", addr, err)
			continue
		} 
		s.logger.Infof("gRPC server listening on TCP address: %s", addr)
		listeners = append(listeners, l)
	}
	return listeners, nil
}
```

Reply:
I have reverted the changes to server.go and server_test.go so:

1. Still tests the creation of the TCP sockets per @mukundansundar 's concerns
2. Does not have this `if` anymore